### PR TITLE
DataLad containers-run (and a bit on DataLad extensions)

### PR DESCRIPTION
**Pull request by adswa**
This section is on DataLad's extensions, and focuses specifically on ``datalad-containers``. 
I have added an overview of existing extensions in the section "Further options". I have included a section on datalad containers in the chapter after the yoda principles/yoda project. 

TODOs:
- [x] check which extensions should be mentioned in the books, get feedback on whether any particular extensions should be covered in more detail (i.e. standalone section) as well
- [x] write a singularity image for demonstration (just needs seaborn, pandas, sklearn), and make it publicly available
- [x] write section that introduces/demonstrates datalad containers run
- [x] write a short summary?

### CODE REVIEW

**Comment by adswa** [comment]
This could work. I've written a basic Singularity image [here](https://github.com/datalad-handbook/resources). If rebased onto the top of the branch ``python`` (#226), this section allows to rerun the analysis performed in there within the container with datalad containers-run.

**Comment by adswa** [comment]
hey @mih, this in still WIP, but take a look at [docs/basics/101-133-containersrun.rst](https://github.com/datalad-handbook/book/blob/ab9f697aca7e960472c9ebe722aacd9ff06c83f5/docs/basics/101-133-containersrun.rst) if you want

File: docs/basics/101-133-containersrun.rst [hunk 1]
```diff
@@ -0,0 +1,306 @@
...
+
+The :command:`datalad containers-add` command takes an arbitrary
+name to give to the container, and a path or url to a container image:
+
+.. runrecord:: _examples/DL-101-133-101
+   :language: console
+   :workdir: dl-101/DataLad-101/midterm_project
+
+   # we are in the midterm_project subdataset
+   $ datalad containers-add python --url shub://adswa/resources:1
```
Comment (mih) [inline docs/basics/101-133-containersrun.rst]: I would likely choose a different, less generic name. It might be confusing to need to have Python installed on a system, in order to run Datalad and `datalad containers-add` to ultimately have a an image with Python that we could not execute with `datalad containers-run` without also having Python outside of the container ...

Comment (mih) [reply docs/basics/101-133-containersrun.rst]: Maybe also worth pointing out the source of that image: https://github.com/datalad-handbook/resources

File: docs/basics/101-134-summary.rst [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+
+For one, you will not be surprised if you ever see a subdataset being shown as
+``modified`` by :command:`datalad status`: You now know that if a subdataset
+evolves, it's most recent state needs to be explicitly saved to the superdatasets
+history.
+
+On a different matter, you are now able to capture and share analysis provenance that
+includes the relevant software environment. This does not only make your analyses
+projects automatically reproducible, but automatically *computationally* reproducible -
+you can make sure that your analyses runs on any computer, regardless of the software
```
Comment (mih) [inline docs/basics/101-134-summary.rst]: any computer with singularity (which is not *any* computer)

File: docs/basics/101-133-containersrun.rst [hunk 3]
```diff
@@ -0,0 +1,306 @@
...
+  The instance that you can actually use for your computations, *build* from
+  an image.
+
+**Hub**
+  A storage resource to share and consume images. There is
+  `Singularity-Hub <https://singularity-hub.org/>`_ and
+  `Docker-Hub <https://hub.docker.com/>`_.
+
+Note that as of now, the ``datalad-containers`` extensions only supports
+Singularity images, but support for Docker is being actively developed.
```
[the remark this answers is not in the document]
Comment (adswa) [reply docs/basics/101-133-containersrun.rst]: ahh, thats news to me. I misread this paragraph in the datalad-containers docs:

> At the moment there is built-in support for Singularity images, but other container execution systems can be used together with custom helper scripts. Direct support for Docker is under development.

In this case, I should learn how to use Docker with ``datalad containers-run``, and include it as well (maybe hidden)

File: docs/basics/101-133-containersrun.rst (hunk 1 above)
Comment (adswa) [reply docs/basics/101-133-containersrun.rst]: that's a very good point. Do you have a preference for any of the following, or other ideas?

- ``iris_env``
- ``midterm-software``

File: docs/basics/101-133-containersrun.rst [hunk 4]
```diff
@@ -0,0 +1,306 @@
...
+(HPC) infrastructure [#f2]_. An alternative is `Singularity <https://sylabs
+.io/docs/>`_.
+Both of these tools share core terminology:
+
+**Image**
+   A template or "recipe" from which containers are build. It lists all
+   required components of the computational environment, and is made by a
+   human user. If you want to create your own container, you start by writing
+   an Image file, but you can also *pull* a publicly shared image from the
+   *Hub* of the tool you are using. Based on images, containers are *build*.
```
[the remark this answers is not in the document]
Comment (adswa) [reply docs/basics/101-133-containersrun.rst]: thx for the clarification!

File: docs/basics/101-133-containersrun.rst (hunk 3 above)
[the remark this answers is not in the document]
Comment (kyleam) [reply docs/basics/101-133-containersrun.rst]: @mih
> I think Docker works fine. @kyleam do you think otherwise?

There are known rough edges and probably a good deal of unknown ones given that the docker adapter hasn't seen heavy use, but the basic idea is there and works at least for simple cases.

@adswa
>  I misread this paragraph in the datalad-containers docs:

No, I don't think you misread it.  That's just stale.  It was added in f9cecf2 (DOC: Basic intro/demo (fixes gh-4), 2018-05-28).  The docker adapter wasn't in the tree then:

```
$ git ls-tree -r f9cecf2 | grep -c docker
0
```

I'll update that text.

[the remark this answers is not in the document]
Comment (adswa) [reply docs/basics/101-133-containersrun.rst]: Thx, @kyleam!

**Comment by adswa** [comment]
Alrighty, I think I have addressed the comments - thx much for them, @mih!

File: docs/basics/101-133-containersrun.rst [hunk 5]
```diff
@@ -0,0 +1,320 @@
...
+(HPC) infrastructure [#f2]_. An alternative is `Singularity <https://sylabs
+.io/docs/>`_.
+Both of these tools share core terminology:
+
+**Recipe**
+   A text file template that lists all required components of the computational environment.
+   It is made by a human user.
+
+**Image**
+   This is *build* from the recipe file. It is a static filesystem inside a file,
```
Comment (kyleam) [inline docs/basics/101-133-containersrun.rst]: drive-by typo comment: s/build/built/

Comment (adswa) [reply docs/basics/101-133-containersrun.rst]: thx!

Comment (adswa) [inline docs/basics/101-133-containersrun.rst]:
```suggestion
   This is *built* from the recipe file. It is a static filesystem inside a file,
```
